go-controller/pkg/ovn/egressip.go

```diff
@@ -416,55 +416,94 @@ func (e *egressIPMode) needsRetry(pod *kapi.Pod) bool {
-				return fmt.Errorf("unable to create logical router policy for egress IP: %s, stderr: %s, err: %v", status.EgressIP, stderr, err)
-			}
+			filterOption = fmt.Sprintf("ip4.src == %s && ip4.dst == 0.0.0.0/0", podIP.String())
+		}
+		if packetMark != 0 {
+			_, stderr, err = util.RunOVNNbctl(
+				"--id=@lr-policy",
+				"create",
+				"logical_router_policy",
+				"action=reroute",
+				fmt.Sprintf("match=\"%s\"", filterOption),
+				fmt.Sprintf("priority=%v", egressIPRereoutePriority),
+				fmt.Sprintf("nexthop=%s", gatewayRouterIP),
+				fmt.Sprintf("external_ids:name=%s", egressIPName),
+				fmt.Sprintf("pkt_mark=%v", packetMark),
+				"--",
+				"add",
+				"logical_router",
+				ovnClusterRouter,
+				"policies",
+				"@lr-policy",
+			)
+		} else {
+			_, stderr, err = util.RunOVNNbctl(
+				"--id=@lr-policy",
+				"create",
+				"logical_router_policy",
+				"action=reroute",
+				fmt.Sprintf("match=\"%s\"", filterOption),
+				fmt.Sprintf("priority=%v", egressIPRereoutePriority),
+				fmt.Sprintf("nexthop=%s", gatewayRouterIP),
+				fmt.Sprintf("external_ids:name=%s", egressIPName),
+				"--",
+				"add",
+				"logical_router",
+				ovnClusterRouter,
+				"policies",
+				"@lr-policy",
+			)
+		}
+		if err != nil {
+			return fmt.Errorf("unable to create logical router policy: %s, stderr: %s, err: %v", status.EgressIP, stderr, err)
 		}
 	}
 	return nil
 }
 
-func (e *egressIPMode) deleteEgressPolicy(podIps []net.IP, status egressipv1.EgressIPStatusItem) error {
+func (e *egressIPMode) deleteEgressPolicy(podIps []net.IP, status egressipv1.EgressIPStatusItem, egressIPName string) error {
 	for _, podIP := range podIps {
+		var filterOption string
 		if utilnet.IsIPv6(podIP) && utilnet.IsIPv6String(status.EgressIP) {
-			_, _, err := util.RunOVNNbctl("lr-policy-del", ovnClusterRouter, egressIPRereoutePriority, fmt.Sprintf("ip6.src == %s && ip6.dst == ::/0", podIP.String()))
-			if err != nil {
-				return fmt.Errorf("unable to delete logical router policy for pod IP: %s, err: %v", podIP.String(), err)
-			}
+			filterOption = fmt.Sprintf("ip6.src == %s && ip6.dst == ::/0", podIP.String())
 		} else if !utilnet.IsIPv6(podIP) && !utilnet.IsIPv6String(status.EgressIP) {
-			_, _, err := util.RunOVNNbctl("lr-policy-del", ovnClusterRouter, egressIPRereoutePriority, fmt.Sprintf("ip4.src == %s && ip4.dst == 0.0.0.0/0", podIP.String()))
+			filterOption = fmt.Sprintf("ip4.src == %s && ip4.dst == 0.0.0.0/0", podIP.String())
+		}
+		policyIDs, stderr, err := util.RunOVNNbctl(
+			"--data=bare",
+			"--no-heading",
+			"--columns=_uuid",
+			"find",
+			"logical_router_policy",
+			fmt.Sprintf("match=\"%s\"", filterOption),
+			fmt.Sprintf("priority=%v", egressIPRereoutePriority),
+			fmt.Sprintf("external_ids:name=%s", egressIPName),
+		)
+		if err != nil {
+			return fmt.Errorf("unable to find logical router policy: %s, stderr: %s, err: %v", status.EgressIP, stderr, err)
+		}
+		for _, policyID := range strings.Split(policyIDs, "\n") {
```

Comment:
If there could actually be multiple rows then you need to split by `"\n\n"`, not `"\n"`. (Because that's how `ovn-*ctl find` works.) But there can only be one row output here anyway, right?

Reply:
The `create`/`add` commands do not fail if the command is duplicated. So on restart we might risk adding the same policy twice, which is why I added this. It might be better checking that we never add something that already exists, but I figured that makes add one OVN transaction more costly.

Reply:
FYI: even though the `create`/`add` commands lead to duplicated OVN objects, they do not seem to result in duplicated openflows:

```
ovn-nbctl --db tcp:172.21.0.2:6641 lr-policy-list ovn_cluster_router
Routing Policies
      1005 ip4.src == 10.244.0.2 && ip4.dst == 172.21.0.3 /* ovn-worker */         reroute               169.254.0.1
      1005 ip4.src == 10.244.1.2 && ip4.dst == 172.21.0.2 /* ovn-control-plane */         reroute               169.254.0.1
      1005 ip4.src == 10.244.2.2 && ip4.dst == 172.21.0.4 /* ovn-worker2 */         reroute               169.254.0.1
      1004 inport == "rtos-ovn-control-plane" && ip4.dst == 172.21.0.2 /* ovn-control-plane */         reroute                10.244.1.2
      1004 inport == "rtos-ovn-worker" && ip4.dst == 172.21.0.3 /* ovn-worker */         reroute                10.244.0.2
      1004 inport == "rtos-ovn-worker2" && ip4.dst == 172.21.0.4 /* ovn-worker2 */         reroute                10.244.2.2
       101 ip4.src == 10.244.0.0/16 && ip4.dst == 10.244.0.0/16           allow
       101 ip4.src == 10.244.0.0/16 && ip4.dst == 172.21.0.2/32           allow
       101 ip4.src == 10.244.0.0/16 && ip4.dst == 172.21.0.3/32           allow
       101 ip4.src == 10.244.0.0/16 && ip4.dst == 172.21.0.4/32           allow
       100      ip4.src == 10.244.0.3 && ip4.dst == 0.0.0.0/0         reroute                100.64.2.1
       100      ip4.src == 10.244.0.3 && ip4.dst == 0.0.0.0/0         reroute                100.64.1.1
       100      ip4.src == 10.244.0.3 && ip4.dst == 0.0.0.0/0         reroute                100.64.1.1
       100      ip4.src == 10.244.2.3 && ip4.dst == 0.0.0.0/0         reroute                100.64.1.1

[root@ovn-control-plane ~]# ovs-ofctl -O OpenFlow13 dump-flows br-int | grep 10.244.0.3
 cookie=0xa421b605, duration=198.260s, table=20, n_packets=0, n_bytes=0, priority=100,ip,metadata=0x1,nw_src=10.244.0.3 actions=load:0x64400101->NXM_NX_XXREG0[96..127],load:0x64400102->NXM_NX_XXREG0[64..95],set_field:0a:58:64:40:01:02->eth_src,set_field:0x7->reg15,load:0x1->NXM_NX_REG10[0],resubmit(,21)
 cookie=0x3cd483df, duration=22978.596s, table=21, n_packets=0, n_bytes=0, priority=50,arp,metadata=0x4,arp_tpa=10.244.0.3,arp_op=1 actions=move:NXM_OF_ETH_SRC[]->NXM_OF_ETH_DST[],set_field:0a:58:0a:f4:00:03->eth_src,set_field:2->arp_op,move:NXM_NX_ARP_SHA[]->NXM_NX_ARP_THA[],set_field:0a:58:0a:f4:00:03->arp_sha,move:NXM_OF_ARP_SPA[]->NXM_OF_ARP_TPA[],set_field:10.244.0.3->arp_spa,move:NXM_NX_REG14[]->NXM_NX_REG15[],load:0x1->NXM_NX_REG10[0],resubmit(,32)

``` 

I am thus unsure if it's worth running a `find` command before each `create` command, when setting up egress for pods.

Reply:
OK, circling around on this: I am adding `--format=csv` to avoid the `\n\n`. Concerning the duplicated logical router policies / nat rules on restart. I don't see any duplicated openflows, even though the OVN objects are duplicated. I will need to confirm this tomorrow with @numansiddique , i.e: that having the exact same `Logical_router_policy` and `NAT` object twice does not duplicate the openflows.

Reply:
> I don't see any duplicated openflows, even though the OVN objects are duplicated.

Yeah, it's a no-op to re-add a flow that already exists. (It neither duplicates the flow nor returns an error.)

```
[root@p50 ovn-kubernetes]# ovs-ofctl add-flow br0 table=0,priority=100,ip,actions=output:0
[root@p50 ovn-kubernetes]# ovs-ofctl add-flow br0 table=0,priority=100,ip,actions=output:0
[root@p50 ovn-kubernetes]# echo $?
0
[root@p50 ovn-kubernetes]# ovs-ofctl dump-flows br0
 cookie=0x0, duration=1.334s, table=0, n_packets=0, n_bytes=0, priority=100,ip actions=output:0
[root@p50 ovn-kubernetes]# 
```

Reply:
So are you fine with not performing an additional `find` before doing a `create`, and having the delete remove multiple IDs (should they exist), @danwinship ?

Reply:
> The `create`/`add` commands do not fail if the command is duplicated. So on restart we might risk adding the same policy twice, which is why I added this.

Huh? This is the delete code, not the add code...

We should not re-add the same OVN object on every restart, even if it doesn't result in duplicate OpenFlow. If we have to do a `find`-then-`create` to ensure that, then we should do that

Reply:
> Huh? This is the delete code, not the add code...

Yeah, I was explaining why the delete code looked the way it did, which was because of the add code. 

I will make sure we `find`-then-`create` then.